File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.py
class Solution(object):
    def fun(self,bloomDay,mid,m,k):
        c=0
        bouquets=0
        for i in bloomDay:
            if(i<=mid):
                c+=1
                if(c==k):
                    bouquets+=(c//k)
                    c=0
            else:
                c=0
        if(bouquets>=m):
            return True
        else:
            return False 
    def minDays(self, bloomDay, m, k):
        """
        :type bloomDay: List[int]
        :type m: int
        :type k: int
        :rtype: int
        """           
        n=len(bloomDay)
        if(m*k>n):
            return -1
        low=min(bloomDay)
        high=max(bloomDay)
        ans=-1
        while(low<=high):
            mid=(low+high)//2
            total=self.fun(bloomDay,mid,m,k)
            if(total):
                ans=mid
                high=mid-1
            else:
                low=mid+1
        return ans            

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        if(m*k>bloomDay.length)return -1;
        int low=Integer.MAX_VALUE,high=Integer.MIN_VALUE;
        for(int i: bloomDay){
            low=Math.min(low,i);
            high=Math.max(high,i);
        }
        int ans=-1;
        while(low<=high){
            int mid=(low+high)/2;
            int b=0,c=0;
            for(int i:bloomDay){
                if(i<=mid){
                    c+=1;
                    if(c==k){
                        b+=1;
                        c=0;
                    }
                }
                else c=0;
            }
            if(b>=m){
                ans=mid;
                high=mid-1;
            }
            else low=mid+1;
        }
        return ans;
    }
}